lib/handler/file.c

```diff
@@ -110,6 +119,46 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
         do_close(&self->super, req);
 }
 
+static void do_multirange_proceed(h2o_generator_t *_self, h2o_req_t *req) {
+    struct st_h2o_sendfile_generator_t *self = (void *)_self;
+    size_t rlen, *range_cur = self->range_infos + 2 * self->finished_range, used_buf = 0;
+    ssize_t rret;
+    h2o_iovec_t vec;
+
+    if (self->bytesleft == 28) {
+        vec.len = sprintf(self->buf, "\r\n--%s--\r\n", self->boundary.base);
+        vec.base = self->buf;
+        h2o_send(req, &vec, 1, 1);
+        do_close(&self->super, req);
+        return;
```

Comment:
> TBO I don't fully understand how `do_procceed()` work when file is larger than `buf`, though I just followed that function and write `do_multirange_proceed()`. What is the process after non-finished sending, i.e. when leave the proceed function after calling h2o_send with `is_finished` 0? Sorry for not fully understanding the workflow, I tried my best...

There is nothing for you to apologize.  I highly appreciate your contribution!

When `is_finished` is set to zero, the protocol handlers call the `proceed` callback of the generator when data passed in via `h2o_send` has been sent to the client.  This phase is repeated until the generator calls `h2o_send` with `is_finished` set to 1.

In other words, you should change the value of `generator->super.proceed` from `do_proceed` to `do_multipart_proceed` in case of a multi-part response (sorry I did not notice the issue when I read your code).

Regarding the issue bet. push and pull generators, the former is the default one which should be supported by all generators.  Pull generator is an _optimization_ which can be used only when the underlying procotol handlers / output streams provide support for the feature.  For example, [the chunked encoder only provides the pull interface](https://github.com/h2o/h2o/blob/00c44dbb3eb49a73d6222a8df77090804b736a3b/lib/handler/chunked.c#L89) (notice that only the `do_send` callback is registered, `start_pull` remains NULL).

So for the PR to work correctly you must implement the pushing generator appropriately, whereas the pulling generator being something that can be worked on as an improvement.  That is why I suggested concentrating on the pushing generator for the time being.